legion/engine/llri/llri.hpp

```diff
@@ -0,0 +1,375 @@
+            case Result::ErrorFeatureNotSupported:
+                return "ErrorFeatureNotSupported";
+            case Result::ErrorExtensionNotSupported:
+                return "ErrorExtensionNotSupported";
+            case Result::ErrorDeviceHung:
+                return "ErrorDeviceHung";
+            case Result::ErrorDeviceLost:
+                return "ErrorDeviceLost";
+            case Result::ErrorDeviceRemoved:
+                return "ErrorDeviceRemoved";
+            case Result::ErrorDriverFailure:
+                return "ErrorDriverFailure";
+            case Result::NotReady:
+                return "NotReady";
+            case Result::ErrorOutOfHostMemory:
+                return "ErrorOutOfHostMemory";
+            case Result::ErrorOutOfDeviceMemory:
+                return "ErrorOutOfDeviceMemory";
+            case Result::ErrorInitializationFailed:
+                return "ErrorInitializationFailed";
+            case Result::ErrorIncompatibleDriver:
+                return "ErrorIncompatibleDriver";
+        }
+
+        return "Invalid Result value";
+    }
+
+    /**
+     * @brief Instance description to be used in llri::createInstance().
+    */
+    struct InstanceDesc
+    {
+        /**
+         * @brief The number of instance extensions in the InstanceDesc::extensions array.
+        */
+        uint32_t numExtensions;
+        /**
+         * @brief The instance extensions, if InstanceDesc::numExtensions > 0, then this has to be a valid pointer to an array of InstanceExtension.
+         * If numExtensions == 0, then this pointer may be nullptr.
+        */
+        InstanceExtension* extensions;
+        /**
+         * @brief Sets the name of the application in internal API if applicable.
+         * This is not guaranteed but is known to at least apply to Vulkan.
+        */
+        const char* applicationName;
+
+        InstanceDesc() = default;
+        InstanceDesc(const uint32_t& numExtensions, InstanceExtension* extensions, const char* applicationName) : numExtensions(numExtensions), extensions(extensions), applicationName(applicationName) { }
+    };
+
+    /**
+     * @brief Create an llri Instance.
+     * Like with all API objects, the user is responsible for destroying the instance again using destroyInstance().
+     * @param desc The description of the instance.
+     * @param instance Must be a valid pointer to an Instance variable.
+     *
+     * @return Success upon correct execution of the operation.
+     * @return ErrorInvalidUsage if the instance is nullptr, or if desc.numExtensions > 0 and desc.extensions is nullptr.
+     * @return ErrorExtensionNotSupported if any of the extensions fail to be created.
+     * @return Otherwise it may also return: ErrorExtensionNotSupported, ErrorOutOfHostMemory, ErrorOutOfDeviceMemory, ErrorInitializationFailed, ErrorIncompatibleDriver.
+    */
+    Result createInstance(const InstanceDesc& desc, Instance* instance);
```

Comment:
Since Instance is an alias for InstanceT*, that'd make it an Instance*&, are we okay with that? I figured using pointers for out parameters would be a good/clean convention, but I can switch to references if that'd be preferred

Reply:
I mean a reference also clearly communicates read-write usage. a const ref communicates read-only usage. it's not out of line with legion standard, just not necessarily modern C++.

Reply:
@LeonBrands